pupil_src/shared_modules/calibration_routines/optimization_calibration/calibrate_3d.py

```diff
@@ -0,0 +1,218 @@
+    # Minimize reprojection-type errors by moving the 3d gaze targets and
+    # adjusting the orientation of the eyes while fixing their positions.
+
+    # Find initial guess for the poses in world coordinates
+    initial_rotation0 = utils.get_initial_eye_camera_rotation(
+        pupil0_normals, unprojected_ref_points
+    )
+    initial_rotation1 = utils.get_initial_eye_camera_rotation(
+        pupil1_normals, unprojected_ref_points
+    )
+    initial_translation0 = eye0_hardcoded_translation
+    initial_translation1 = eye1_hardcoded_translation
+
+    # world cam and eyes are viewed as spherical cameras of unit radius
+    world = SphericalCamera(
+        observations=unprojected_ref_points,
+        rotation=np.zeros(3),
+        translation=np.zeros(3),
+        fix_rotation=True,
+        fix_translation=True,
+    )
+    eye0 = SphericalCamera(
+        observations=pupil0_normals,
+        rotation=initial_rotation0,
+        translation=initial_translation0,
+        fix_rotation=False,
+        fix_translation=True,
+    )
+    eye1 = SphericalCamera(
+        observations=pupil1_normals,
+        rotation=initial_rotation1,
+        translation=initial_translation1,
+        fix_rotation=False,
+        fix_translation=True,
+    )
+
+    initial_spherical_cameras = world, eye0, eye1
+    initial_gaze_targets = unprojected_ref_points * initial_depth
+
+    ba = bundle_adjustment.BundleAdjustment(fix_gaze_targets=False)
+    residual, poses_in_world, gaze_targets_in_world = ba.calculate(
+        initial_spherical_cameras, initial_gaze_targets
+    )
+
+    success = residual < residual_threshold
+    return success, poses_in_world, gaze_targets_in_world
+
+
+def calibrate_monocular(unprojected_ref_points, pupil_normals, pupil_id, initial_depth):
+    """Determine the poses of the eyes and 3d gaze points by solving a specific
+    least-squares minimization
+
+    :param unprojected_ref_points: the unprojection of the observed 2d reference points
+    at unit distance in world camera coordinates
+    :param pupil_normals: eye's pupil normals in eye camera coordinates
+    :param pupil_id: pupil id (0 or 1)
+    :param initial_depth: initial guess of the depth of the gaze targets
+    :return: optimized poses and 3d gaze targets in world camera coordinates
+    """
+    # Monocular calibration strategy:
+    # Take eye as the origin and express everything in eye coordinates.
+    # Minimize reprojection-type errors by moving world cam
+    # while fixing the 3d gaze targets.
+
+    # Find initial guess for the poses in eye coordinates
+    initial_rotation_matrix, _ = utils.find_rigid_transform(
+        unprojected_ref_points, pupil_normals
+    )
+    hardcoded_translation = (
+        eye0_hardcoded_translation if pupil_id == 0 else eye1_hardcoded_translation
+    )
+    initial_rotation = cv2.Rodrigues(initial_rotation_matrix)[0].ravel()
+    initial_translation = -np.dot(initial_rotation_matrix, hardcoded_translation)
```

Comment:
@ChingT Why is the dot product applied to this initial translation, but not in the binocular case? Has this something to do with the fact that you pass (0, 0, 0) below?